lib/sshutils/server.go

```diff
@@ -841,17 +842,31 @@ func (c *connectionWrapper) Read(b []byte) (int, error) {
 	if bytes.HasPrefix(buff, multiplexer.ProxyV2Prefix) {
 		reader := bufio.NewReader(io.MultiReader(bytes.NewBuffer(buff), c.Conn))
 
-		newProxyLine, err := multiplexer.ReadProxyLineV2(reader)
+		proxyLine, err := multiplexer.ReadProxyLineV2(reader)
 		if err != nil {
 			return 0, trace.Wrap(err)
 		}
-		if newProxyLine.IsSigned() && c.caGetter != nil {
-			err = newProxyLine.VerifySignature(context.Background(), c.caGetter, c.clusterName, c.clock)
+		if proxyLine != nil && proxyLine.IsSigned() && c.caGetter != nil {
+			err = proxyLine.VerifySignature(context.Background(), c.caGetter, c.clusterName, c.clock)
+			// NOTE(anton): Temporarily using string comparison here to not create circular references.
+			// Will be refactored after #21835 is resolved.
 			if err != nil {
-				return 0, trace.Wrap(err)
+				if strings.Contains(err.Error(), "could not get specified host CA to verify signed PROXY header") {
```

Comment:
Can't `VerifySignature` return different errors for each case?

would simplify a lot if it's possible to:

```go
switch {
case errors.Is(err, somepack.ErrHostCANotFound):
	c.logger.WithFields(logrus.Fields{
					"src_addr": c.Conn.RemoteAddr(),
					"dst_addr": c.Conn.LocalAddr(),
				}).Warn("Could not verify PROXY signature for connection - could not get host CA")
case errors.Is(err, somepack.ErrUnknownSigningCert):
c.logger.WithFields(logrus.Fields{
					"src_addr": c.Conn.RemoteAddr(),
					"dst_addr": c.Conn.LocalAddr(),
				}).Warn("Could not verify PROXY signature for connection - signed by non local cluster")
case err!=nil:
return 0, trace.Wrap(err)
case proxyLine.IsVerified:
c.clientAddr = &proxyLine.Source
}

```

You can later refactor the code and move the errs into another package to fix #21835

Remove the `f` from the `Warnf` if you won't pass any formating

Reply:
`VerifySignature` already does return different errors and they are used in multiplexer. I just want to leave handling this during #21835, because it will probably change (and simplify) this function more significantly than just handling errors.

Remove `Warnf()` 👍